cpp/src/arrow/compute/kernels/vector_run_end_encode.cc

```diff
@@ -254,7 +254,7 @@ struct RunEndEncodeExec {
       return RunEndEncodeNullArray(TypeTraits<RunEndType>::type_singleton(), ctx,
                                    input_array, result);
     } else {
-      const bool has_validity_buffer = input_array.MayHaveNulls();
+      const bool has_validity_buffer = input_array.GetNullCount() > 0;
```

Comment:
@mapleFU this is the fix to the issue. `MayHaveNulls` was returning `true` here because the validity bitmap exists, but down the road, when the REE is allocated, validity buffer was assumed to also exist in the allocated output. When, in fact, it didn't actually exist because `null_count == 0` gets that buffer cleared.